log-replication/src/main/java/org/corfudb/logreplication/receiver/LogEntryWriter.java

```diff
@@ -1,80 +1,136 @@
+                    MessageType.LOG_ENTRY_MESSAGE, srcGlobalSnapshot);
+            throw new Exception("wrong type of message");
         }
     }
 
     /**
-     * find the msg whose
-     * @param preAddress
-     * @return
+     * Convert message data to an MultiObjectSMREntry and write to log.
+     * @param txMessage
      */
-    TxMessage findMsg(long preAddress) {
-        //delete all message whose address < preAddress
-        //return the msg whose preAddress == preAddress
-        return new TxMessage();
-    }
+    void processMsg(TxMessage txMessage) {
+        OpaqueEntry opaqueEntry = OpaqueEntry.deserialize(Unpooled.wrappedBuffer(txMessage.getData()));
+        Map<UUID, List<SMREntry>> map = opaqueEntry.getEntries();
+        if (!streamUUIDs.contains(map.keySet())) {
+            log.error("txMessage contains noisy streams {}, expecting {}", map.keySet(), streamUUIDs);
+        }
+        try {
+            rt.getObjectsView().TXBegin();
+            TokenResponse tokenResponse = rt.getSequencerView().next((UUID[])(map.keySet().toArray()));
+            MultiObjectSMREntry multiObjectSMREntry = new MultiObjectSMREntry();
+            for (UUID uuid : opaqueEntry.getEntries().keySet()) {
+                for(SMREntry smrEntry : opaqueEntry.getEntries().get(uuid)) {
+                    multiObjectSMREntry.addTo(uuid, smrEntry);
+                }
+            }
+            rt.getAddressSpaceView().write(tokenResponse.getToken(), multiObjectSMREntry);
 
-    void setContext(long snapshot) {
-        srcGlobalSnapshot = snapshot;
-        lastSrcAddressProcessed = snapshot;
+        } finally {
+            rt.getObjectsView().TXEnd();
+        }
     }
 
-    //Question? what the size of the deltaQue? Is it a in memory que or persistent que?
-    void processDeltaQue() {
+    /**
+     * Go over the queue, if the next expecting msg is in queue, process it.
+     * @throws Exception
+     */
+    void processQue() throws Exception {
         while (true) {
-            TxMessage msg = findMsg(lastSrcAddressProcessed);
-            if (msg == null) {
+            TxMessage txMessage = msgQ.get(lastMsgTs);
+            if (txMessage == null) {
                 return;
             }
+            processMsg(txMessage);
+            msgQ.remove(lastMsgTs);
+            lastMsgTs = txMessage.getMetadata().getEntryTimeStamp();
+        }
+    }
+
+    void applyTxMessage(TxMessage msg) throws Exception {
+        verifyMetadata(msg.getMetadata());
+
+        // Ignore the out of date messages
+        if (msg.getMetadata().getSnapshotTimestamp() < srcGlobalSnapshot) {
+            log.warn("Received message with snapshot {} is smaller than current snapshot {}.Ignore it",
+                    msg.getMetadata().getSnapshotTimestamp(), srcGlobalSnapshot);
+            return;
+        }
+
+        // A new full sync happens, setup the new srcGlobalSnapshot
+        if (msg.getMetadata().getSnapshotTimestamp() > srcGlobalSnapshot) {
```

Comment:
when can this happen ? isn't the verify going to throw an exception at the beginning of the method.

Reply:
The leadership change, the new leader does a new snapshot sync and a new delta sync.